fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,139 @@
+     * update state with the received todos.
+     */
+    fetch('/api/todos')
+      .then(res => {
+        if (!res.ok) {
+          throw new Error('Network response was not OK');
+        }
+        return res.json();
+      })
+      .then(data => {
+        this.setState({ todos: data });
+      })
+      .catch(error => console.error('Error:', error));
+  }
+
+  addTodo(newTodo) {
+    /**
+    * Use fetch to send a POST request to `/api/todos`.
+    * Then 😉, once the response JSON is received and parsed,
+    *   - create a shallow copy of the todos array from state
+    *   - add the todo received from the server to the copied array
+    *   - replace the old todos array in state with the new one
+    *
+    * DO NOT MUTATE the original state array, nor any objects within it.
+    * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+    *
+    * TIP: Be sure to SERIALIZE the todo object in the body with JSON.stringify()
+    * and specify the "Content-Type" header as "application/json"
+    *
+    * TIP: Use Array.prototype.concat to create a new array containing the contents
+    * of the old array, plus the object returned by the server.
+    */
+
+    fetch('/api/todos', {
+      method: 'POST',
+      headers: {
+        'Content-type': 'application/json'
+      },
+      body: JSON.stringify(newTodo)
+    })
+      .then(res => {
+        if (!res.ok) {
+          throw new Error('Network response not OK');
+        }
+        return res.json();
+      })
+      .then(data => {
+        const todosArray = this.state.todos;
+        this.setState({ todos: todosArray.concat(data) });
+      })
+      .catch(error => console.error('Error:', error));
+  }
+
+  toggleCompleted(todoId) {
+    /**
+     * Find the index of the todo with the matching todoId in the state array.
+     * Get its "isCompleted" status.
+     * Make a new object containing ONE PROPERTY: the opposite "isCompleted" status.
+     * Use fetch to send a PATCH request to `/api/todos/${todoId}`
+     * Then 😉, once the response JSON is received and parsed,
+     *   - create a shallow copy of the todos array from state
+     *   - replace the old todo with the todo received from the server
+     *   - replace the old todos in the state with the new one (you know the index).
+     *
+     * NOTE: "toggle" means to flip back and forth, so clicking a todo
+     * in the list repeatedly should "toggle" its isCompleted status back and forth.
+     *
+     * DO NOT try to calculate the index of the todo by subtracting 1 from the id.
+     *
+     * DO NOT MUTATE the original state array, nor any objects within it.
+     * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+     *
+     * TIP: Be sure to SERIALIZE the updates in the body with JSON.stringify()
+     * And specify the "Content-Type" header as "application/json"
+     */
+    const todosArray = this.state.todos;
```

Comment:
This won't make a copy of the array, it will just save the memory address of the array into `todosArray` so down below when you make changes to `todosArray` you are mutating the state. This can lead to your app not updating correctly and errors in your data. To make a copy of an array you have a few options, here are 2 of the more common ways:

Using [slice](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice)
```JAVASCRIPT
const todosArray = this.state.todos.slice();
```

Using ES6 [spread operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax)
```JAVASCRIPT
const todosArray = [...this.state.todos];
```